Model backup

delhomer opened the pull request:
Introduce backups in `kerastrain.py`. Training a model begins by getting old backups on the file system, if they exist. Then new backups are stored at the end of each training epoch if they improve the best score. For now the reference metric is the validation accuracy.

Additionally a new test has been added to control the model backup process.

delhomer committed:
tests: add a new test for verifying the neural network backup recovering

When saved on the file system under the `.h5` format, neural network
models can be recovered with `load_weights` method: applied to a
model, it affects the saved weights to the current model object. This
commit add a test function to control this behavior on a small
example.

delhomer committed:
training: introduce model backup recovering into kerastrain.py

From this commit, old saved model weights can be recovered to continue
past training processes. Model weights are directly affected to the
current `model` object, and an iteration counter is set starting from
the backup file counter (*e.g.* `checkpoint-epoch-03.h5` means that
the previous training process was saved after 3 epochs, then the
current counter must be equal to 3).

delhomer committed:
training: fix checkpoint recovering

The `if` case that handles model checkpoint recovering is wrongly
set. Two contradictory logs are printed onto the console, and epoch
counter is always set as 0, even if there is a model backup on the
file system. A `else` statement is added to branch the cases
correctly.

delhomer committed:
training: fix backup filename when recovering a trained model

As backup filename when sorted by epoch, backup were set as the last
item in the backup repository. However, it seems that the `os.listdir`
method does not provide the files in the alphabetic order;
consequently we do not get the most recent backup. By using `max`
function applied on backup filenames, we ensure that the most recent
backup is chosen (*i.e.* the backup with the largest epoch).

garaud commented on deeposlandia/kerastrain.py:
```diff
@@ -210,23 +210,56 @@ def add_training_arguments(parser):
+    if os.path.isdir(checkpoint_path):
+        checkpoints = os.listdir(checkpoint_path)
+        if len(checkpoints) > 0:
+            model_checkpoint = max(checkpoints)
```
This a alphanum order. Will the selected file be always the most recent?

delhomer replied:
As we define the checkpoint with the parameter `save_best_only=True`, and as the checkpoint names depend only on current training epoch, all the saved model are supposed to provide a better validation accuracy than the previous ones. In such a context, considering the most recent checkpoint guarantees recovering the best validation accuracy.

garaud commented on deeposlandia/kerastrain.py:
```diff
@@ -210,23 +210,56 @@ def add_training_arguments(parser):
+                           "The model will be trained from scratch."))
+        trained_model_epoch = 0
+
+    checkpoint_filenames = os.path.join(folders["output"], "checkpoints",
```
I prefer to have a variable name in singular.

delhomer replied:
Yes, write. This is fixed.

garaud commented on deeposlandia/kerastrain.py:
```diff
@@ -210,23 +210,56 @@ def add_training_arguments(parser):
+        trained_model_epoch = 0
+
+    checkpoint_filenames = os.path.join(folders["output"], "checkpoints",
+                                        instance_name, "checkpoint-epoch-{epoch:02d}.h5")
```
just a question: you assume with this `02d` that the number of epochs won't be higher than 1000?

delhomer replied:
It won't even be higher than 100. I assumed indeed than the number of epochs won't be larger than this amount, however this may be a strong assumption if we consider larger models, and/or if we decompose epoches in smaller sets.

I may modify `02d` in `03d` for a sake of care.

*As a complementary remark*, note that this name convention is directly related to your first comment about `max(filename)`.

garaud commented on tests/test_model_training.py:
```diff
@@ -38,6 +38,31 @@ def test_model_training(shapes_image_size, shapes_sample, shapes_sample_config,
+    if os.path.isdir(checkpoint_path):
+        checkpoints = os.listdir(checkpoint_path)
+        if len(checkpoints) > 0:
+            model_checkpoint = checkpoints[-1]
```
It's different from the previous `max(checkpoint_path)` in the implemented function `add_training_arguments`.

delhomer replied:
Yes write, this is a small forget. This is supposed to be `max(checkpoints)`.

garaud commented on tests/test_model_training.py:
```diff
@@ -38,6 +38,31 @@ def test_model_training(shapes_image_size, shapes_sample, shapes_sample_config,
+    assert trained_model_epoch > 0
+    assert len(old_weights) == len(new_weights)
+    assert old_weights[0].shape == new_weights[0].shape
+    assert all(not np.allclose(old_weights[i], new_weights[i]) for i in range(len(new_weights)))
```
I wonder if we can use a numpy testing assert here, such as `np.testing.assert_almost_equal`

delhomer replied:
Why not, I did not know this possibility. I was wondering if the most accurate is really `assert_almost_equal`, as we handle integers. I see a `assert_allclose` which is close from the current used function `np.allclose`.

delhomer replied:
Yark, we do not test if all weights are equivalent. The goal is the opposite one: to check if each new weight is different from each old weight. It seems that there is no convenient numpy test for such a task.

I tried a workaround with a `pytest` context manager, something like:
```
with pytest.raises(AssertionError):
    any(np.testing.assert_allclose(old_weights[i], new_weights[i]) for i in range(len(new_weights)))
```
However it is not clear for me that this snippet is really accurate (*e.g.* the test result is the same if I write `any(not ...)`).

garaud reviewed (changes requested):
Just some comments.

delhomer committed:
training: minor fixes after PR #59 proofreading

delhomer committed:
tests: improve the last backup test definition

Switch from `all` to `any` as we want to ensure that weights are
different in one layer at least. Then use a zip iterator to improve
code readibility, and add a comment for a sake of clarity about the
test objective.
